### PR TITLE
Unbonding queue

### PR DESCRIPTION
**Pull request by jonasW3F**
We are proposing an unbonding queue for Relay Chain tokens to significantly reduce the expected unbonding time. The queue would still maintain sufficient stake accountability to mitigate the profitability of LRAs.

### CODE REVIEW

**Comment by burdges** [comment]
Importantly soundness slashes happen quickly because no-shows cause full escalation within minutes.  It follows these parameter choices need only address the classical long range attacks scenarios.  This wasn't mentioned here.

We'd likely change these parameters if we ever adopted an 80% honest assumption for multiple relay chains, but this could be addressed in future.

**Comment by ggwpez** [comment]
@ank4n @gpestana @kianenigma PTAL

File: text/0092-unbonding_queue.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+## Summary
+
+This RFC proposes a flexible unbonding mechanism for tokens that are locked from [staking](https://wiki.polkadot.network/docs/learn-staking) on the Relay Chain (DOT/KSM), aiming to enhance user convenience without compromising system security. 
+
+Locking tokens for staking ensures that Polkadot is able to slash tokens backing misbehaving validators. With changing the locking period, we still need to make sure that Polkadot can slash enough tokens to deter misbehaviour. This means that not all tokens can be unbonded immediately, however we can still allow some tokens to be unbonded quickly.
+
+The new mechanism leads to a signficantly reduced unbonding time on average, by queuing up new unbonding requests and scaling their unbonding duration relative to the size of the queue. New requests are executed with a minimum of 2 days, when the queue is comparatively empty, to the conventional 28 days, if the sum of requests (in terms of stake) exceed some threshold. In scenarios between these two bounds, the unbonding duration scales proportionately. The new mechanism will never be worse than the current fixed 28 days.
```
Comment (kianenigma) [inline text/0092-unbonding_queue.md:15]: > if the sum of requests (in terms of stake) exceed some threshold.

dq: Which part of your formulas below corresponds to this? I fail to detect it.

Comment (jonasW3F) [reply text/0092-unbonding_queue.md:15]: The parameter is called `max_unstake`. Further below it reads:

> We also store a variable, max_unstake that tracks how much stake we allow to unbond potentially earlier than 28 eras (28 days on Polkadot and 7 days on Kusama). 

The queue scales proportionally between 2 and 28 days (with respect to `max_unstake`). In case we exceed that value, the unbonding time is capped at 28 days.

File: text/0092-unbonding_queue.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+Note: Our proposition solely focuses on the locks imposed from staking. Other locks, such as governance, remain unchanged. Also, this mechanism should not be confused with the already existing feature of [FastUnstake](https://wiki.polkadot.network/docs/learn-staking#fast-unstake), which lets users unstake tokens immediately that have not received rewards for 28 days or longer.
+
+As an initial step to gauge its effectiveness and stability, it is recommended to implement and test this model on Kusama before considering its integration into Polkadot, with appropriate adjustments to the parameters. In the following, however, we limit our discussion to Polkadot.
+
+## Motivation
+
+Polkadot has one of the longest unbonding periods among all Proof-of-Stake protocols, because security was the most important goal. Staking on Polkadot is still attractive compared to other protocols because of its above-average staking APY. However the long unbonding period harms usability and deters potential participants that want to contribute to the security of the network.
```
Comment (gpestana) [inline text/0092-unbonding_queue.md:25]:
```suggestion
Polkadot has one of the longest unbonding periods among all Proof-of-Stake protocols, because security is the most important goal. Staking on Polkadot is still attractive compared to other protocols because of its above-average staking APY. However the long unbonding period harms usability and deters potential participants that want to contribute to the security of the network. 
```

File: text/0092-unbonding_queue.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+## Explanation
+
+Before diving into the details of how to implement the unbonding queue, we give readers context about why Polkadot has a 28-day unbonding period in the first place. The reason for it is to prevent long-range attacks (LRA) that becomes theoretically possible if more than 1/3 of validators collude. In essence, a LRA describes the inability of users, who disconnect from the consensus at time t0 and reconnects later, to realize that validators which were legitimate at a certain time, say t0 but dropped out in the meantime, are not to be trusted anymore. That means, for example, a user syncing the state could be fooled by trusting validators that fell outside the active set of validators after t0, and are building a competitive and malicious chain (fork). 
+
+LRAs of longer than 28 days are mitigated by the use of trusted checkpoints, which are assumed to be no more than 28 days old. A new node that syncs Polkadot will start at the checkpoint and look for proofs of finality of later blocks, signed by 2/3 of the validators. In an LRA fork, some of the validator sets may be different but only if 2/3 of  some validator set in the last 28 days signed something incorrect.
```
Comment (gpestana) [inline text/0092-unbonding_queue.md:42]:
```suggestion
LRAs of longer than 28 days are mitigated by the use of trusted checkpoints, which are assumed to be no more than 28 days old. A new node that syncs Polkadot will start at the checkpoint and look for proofs of finality of later blocks, signed by 2/3 of the validators. In an LRA fork, some of the validator sets may be different but only if 2/3 of some validator set in the last 28 days signed something incorrect. 
```

File: text/0092-unbonding_queue.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+LRAs of longer than 28 days are mitigated by the use of trusted checkpoints, which are assumed to be no more than 28 days old. A new node that syncs Polkadot will start at the checkpoint and look for proofs of finality of later blocks, signed by 2/3 of the validators. In an LRA fork, some of the validator sets may be different but only if 2/3 of  some validator set in the last 28 days signed something incorrect. 
+
+If we detect an LRA of no more than 28 days with the current unbonding period, then we should be able to detect misbehaviour from over 1/3 of validators whose nominators are still bonded. The stake backing these validators is considerable fraction of the total stake (empirically it is 0.287 or so). If we allowed more than this stake to unbond, without checking who it was backing, then the LRA attack might be free of cost for an attacker. The proposed mechansim allows up to half this stake to unbond with in 28 days. This halves the amount of tokens that can be slashed, but this is still very high in absolute terms. For example, at the time of writing (19.06.2024) this would translate to around 120 millions DOTs.
```
Comment (gpestana) [inline text/0092-unbonding_queue.md:44]:
```suggestion
If we detect an LRA of no more than 28 days with the current unbonding period, then we should be able to detect misbehaviour from over 1/3 of validators whose nominators are still bonded. The stake backing these validators is considerable fraction of the total stake (empirically it is 0.287 or so). If we allowed more than this stake to unbond, without checking who it was backing, then the LRA attack might be free of cost for an attacker. The proposed mechansim allows up to half this stake to unbond within 28 days. This halves the amount of tokens that can be slashed, but this is still very high in absolute terms. For example, at the time of writing (19.06.2024) this would translate to around 120 millions DOTs.
```

File: text/0092-unbonding_queue.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+Attacks other than an LRA, such as backing incorrect parachain blocks, should be detected and slashed within 2 days. This is why the mechanism has a minimum unbonding period.
```
Comment (gpestana) [inline text/0092-unbonding_queue.md:46]: What are the mechanisms/parameters that ensure that these types of attacks/misbehaviours are always below the 2 day threshold? Could changing some of these params change the threshold to other number of days?

File: text/0092-unbonding_queue.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+```
+unbonding_block_number = min(UPPER_BOUND, max(back_of_unbonding_queue_block_number - current_block_number, LOWER_BOUND)) + current_block_number
+```
+
+Ultimately, the user's token are unbonded at `unbonding_block_number`.
+
+### Proposed Parameters
+There are a few constants to be exogenously set. They are up for discussion, but we make the following recommendation:
+- `MIN_SLASHABLE_SHARE`: `1/2` - This is the share of stake backing the lowest 1/3 of validators that is slashable at any point in time. It offers a trade-off between security and unbonding time. Half is a sensable choice. Here, we have sufficient stake to slash while allowing for a short average unbonding time.
```
Comment (gpestana) [inline text/0092-unbonding_queue.md:83]:
```suggestion
- `MIN_SLASHABLE_SHARE`: `1/2` - This is the share of stake backing the lowest 1/3 of validators that is slashable at any point in time. It offers a trade-off between security and unbonding time. Half is a sensible choice. Here, we have sufficient stake to slash while allowing for a short average unbonding time.
```

File: text/0092-unbonding_queue.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+Owing to the way exposures, which nominators back validators with how many tokens, are stored, it is hard to search for whether a nominator has deferred slashes that need to be applied to them on chain. So we cannot simply check when a nominator attempts to withdraw their bond.
+
+One option would be to allow any account to point out that an unbonding account had a deferred slash and then the chain would  set the `unbonding_block_number` to after the time when the slash would be applied, which will be no more than 28 days from the time the staker unbonded. It is not obvious how to incentivise this, especially in the case that the slash is never applied. Then we would be assuming that in the minimum 2 days unbonding period, not only would any slashable event be caught, but also that someone would post such a transaction cancelling or delaying the unbond until after the slash is applied.
```
Comment (kianenigma) [inline text/0092-unbonding_queue.md:164]: it is either this, which could be fine without incentivization as well. There's also the risk of liveness in such cases, so the transaction must be `Operational`, ensuring an attacker cannot prevent it by filling the chain with remarks for 28 days, it is possibly not super expensive to do that. 

Alternatively, the nominators can be migrated to store their pending slash `on_idle`/`task`. In the few blocks where this `on_idle` is still in progress we disallow any fast unbonds. In honest cases, this is a negligible degradation. It attack cases, the worst, everyone is forced to unbond with 28 days.

**Comment by bhargavbh** [comment]
I would like to point out effects of this RFC on the security of trust-less bridges using Random Sampling (e.g. Snowbridge between Polkadot<>Ethereum). 

**TLDR:** 
With the possibility of faster unbonding, we need to make a stronger assumption on the acceptable downtime of honest relayers for the bridge's safety properties. Earlier, it was sufficient if 1 honest relayer relayed a valid block from Polkadot -> Ethereum within 28 days, which would now reduce to 2 days.  We feel this is still an acceptable period to spin up a new relayer if all the existing relayers are down. 

**Reasoning**
The bridge's security relies on a crypto-economic argument where any validator who BEEFY-signs an un-finalised block gets slashed. Hence, one of the crucial security parameter from bridge security perspective is "how quickly can the stake of the least-backed validator be un-bonded via the queue". Currently, this security parameter is 28 days but introducing the RFC changes it to 2 days.  If all the relayers are down, the a malicious validator can immediately start unbonding and once his stakes are cleared thru the queue (2days), he can attempt relaying un-finalised blocks without any repercussions.

File: text/0092-unbonding_queue.md (hunk 5 above)
Comment (jonasW3F) [reply text/0092-unbonding_queue.md:46]: see [this comment](https://github.com/polkadot-fellows/RFCs/pull/97#issuecomment-2232661059)

**Comment by jonasW3F** [comment]
> I would like to point out effects of this RFC on the security of trust-less bridges using Random Sampling (e.g. Snowbridge between Polkadot<>Ethereum).
> 
> **TLDR:** With the possibility of faster unbonding, we need to make a stronger assumption on the acceptable downtime of honest relayers for the bridge's safety properties. Earlier, it was sufficient if 1 honest relayer relayed a valid block from Polkadot -> Ethereum within 28 days, which would now reduce to 2 days. We feel this is still an acceptable period to spin up a new relayer if all the existing relayers are down.
> 
> **Reasoning** The bridge's security relies on a crypto-economic argument where any validator who BEEFY-signs an un-finalised block gets slashed. Hence, one of the crucial security parameter from bridge security perspective is "how quickly can the stake of the least-backed validator be un-bonded via the queue". Currently, this security parameter is 28 days but introducing the RFC changes it to 2 days. If all the relayers are down, the a malicious validator can immediately start unbonding and once his stakes are cleared thru the queue (2days), he can attempt relaying un-finalised blocks without any repercussions.

Assuming we can relay a valid block from Polkadot to Ethereum within two days is reasonable. Without this block, the bridge would also not be live, causing significant disruptions to the user experience. Furthermore, the stated goal of decentralizing the relayer set greatly reduces the probability of the bridge being non-live for two days.

To ensure reliability, having a backup relayer is crucial. This relayer would not necessarily handle costly transactions between Polkadot and Ethereum but would monitor the main relayer's liveness, stepping in if the main relayer goes offline. Establishing and maintaining this safety net should be a priority. For instance, through initiatives like the Infrastructure Builders Program (Bounty), we have the tools to set up and support this security measure effectively.

**Comment by AlistairStewart** [comment]
2 days is more than enough for the detecting and reporting misbehaviour that is done by validators, for that 1 hour is enough. The question is whether it is enough for things that might require manually running a relayer or bot.

For the bridge, it's ok if all relaying stops, the problem is about not reporting things to Polkadot, and only if we have only malicious relayers running. We should see an attack happening on chain, so this just means that someone should be monitoring the bridged chain, e.g. Ethereum, and have a bot that reports misehaviour, or one that can be run with 2 days notice.

The situation with deferred slashes is similar. We would need someone to be running a bot to detect and report people with deferred slashes unbonding.

And these should be in a state where anyone can spin them up if we have to do so in 2 days, even on a weekend. This is certainly possible.

**Comment by jonasW3F** [comment]
> 2 days is more than enough for the detecting and reporting misbehaviour that is done by validators, for that 1 hour is enough. The question is whether it is enough for things that might require manually running a relayer or bot.
> 
> For the bridge, it's ok if all relaying stops, the problem is about not reporting things to Polkadot, and only if we have only malicious relayers running. We should see an attack happening on chain, so this just means that someone should be monitoring the bridged chain, e.g. Ethereum, and have a bot that reports misehaviour, or one that can be run with 2 days notice.
> 
> The situation with deferred slashes is similar. We would need someone to be running a bot to detect and report people with deferred slashes unbonding.
> 
> And these should be in a state where anyone can spin them up if we have to do so in 2 days, even on a weekend. This is certainly possible.

It sounds like running, maintaining and incentivizing these bots perfectly fall into the scope of the Infrastructure Builders Program (@tugytur)

**Comment by Tomen** [comment]
This is PR has id 97. Therefore this should be RFC-97 NOT 92

**Comment by rzadp** [comment]
> This is PR has id 97. Therefore this should be RFC-97 (NOT 92)

The RFC numbers are not taken from PR ids, because not every PR is an RFC PR, so it would leave unnecessary gaps.

**Comment by ggwpez** [comment]
> The RFC numbers are not taken from PR ids, because not every PR is an RFC PR, so it would leave unnecessary gaps.

They are, as per [README](https://github.com/polkadot-fellows/RFCs?tab=readme-ov-file#process):  
> Rename the file to correspond to the GitHub pull request number and update the "RFC PR" field in the file.

**Comment by jonasW3F** [comment]
> > This is PR has id 97. Therefore this should be RFC-97 (NOT 92)
> 
> The RFC numbers are not taken from PR ids, because not every PR is an RFC PR, so it would leave unnecessary gaps.

That might have been my thinking at the time. I changed the files. When does the [site](https://polkadot-fellows.github.io/RFCs/proposed/0092-unbonding_queue.html) get rebuild? It hasn't updated for a while just checking that there is no issue with caching.

**Comment by rzadp** [comment]
Right, my mistake - I remembered it wrong. It's just that this part of the process is not automatically enforced at any point.
Would a little CI check be helpful, to remind authors if the number is wrong?

@jonasW3F The site gets rebuild nightly.

**Comment by bhargavbh** [comment]
a recent paper on [optimizing unbonding queues for PoS blockchains](https://arxiv.org/pdf/2406.05124) is very relevant work for this RFC. It kind of formally argues why the FCFS mechanism is optimal when the value of withdrawal for agents is homogeneous, backing the ideas presented in this RFC.
@jonasW3F

File: text/0097-unbonding_queue.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+If we detect an LRA of no more than 28 days with the current unbonding period, then we should be able to detect misbehaviour from over 1/3 of validators whose nominators are still bonded. The stake backing these validators is considerable fraction of the total stake (empirically it is 0.287 or so). If we allowed more than this stake to unbond, without checking who it was backing, then the LRA attack might be free of cost for an attacker. The proposed mechansim allows up to half this stake to unbond within 28 days. This halves the amount of tokens that can be slashed, but this is still very high in absolute terms. For example, at the time of writing (19.06.2024) this would translate to around 120 millions DOTs.
+
+Attacks other than an LRA, such as backing incorrect parachain blocks, should be detected and slashed within 2 days. This is why the mechanism has a minimum unbonding period.
+
```
Comment (ordian) [inline text/0097-unbonding_queue.md:47]: we allow for slashing for disputes up to `dispute_period` of past eras, which is configured at 6 days for Polkadot at the moment

File: text/0097-unbonding_queue.md [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+- `Unbonding Days`: The daily expected unbonding days given the history of `daily_unbonded`.
+
+We can observe that historical unbonds only trigger an unbonding time larger than `LOWER_BOUND` in situations with extensive and/or clustered unbonding amounts. The average unbonding time across the whole timeseries is ~2.67 days. We can, however, see it taking effect pushing unbonding times up during large unbonding events. In the largest events, we hit a maximum of 28 days. This gives us reassurance that it is sufficiently sensitive and it makes sense to match the `UPPER_BOUND` with the historically largest unbonds. 
+
+The main parameter affecting the situation is the `max_unstake`. The relationship is obvious: decreasing the `max_unstake` makes the queue more sensitive, i.e., having it spike more quickly and higher with unbonding events. Given that these events historically were mostly associated with parachain auctions, we can assume that, in the absence of major systemic events, users will experience drastically reduced unbonding times.
+The analysis can be reproduced or changed to other parameters using [this repository](https://github.com/jonasW3F/unbonding_queue_analysis).
```
Comment (ordian) [inline text/0097-unbonding_queue.md:108]: the repo seems to be private

File: text/0097-unbonding_queue.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+At any time we store `back_of_unbonding_queue_block_number` which expresses the block number when all the existing unbonders have unbonded.
+
+Let's assume a user wants to unbond some of their stake, i.e., `new_unbonding_stake`, and issues the request at some arbitrary block number denoted as `current_block`. Then:
+
+```
+unbonding_time_delta = new_unbonding_stake / max_unstake * UPPER_BOUND
+```
+
+This number needs to be added to the `back_of_unbonding_queue_block_number` under the conditions that it does not undercut`current_block + LOWER_BOUND` or exceed `current_block + UPPER_BOUND`.
```
Comment (ordian) [inline text/0097-unbonding_queue.md:67]:
```suggestion
This number needs to be added to the `back_of_unbonding_queue_block_number` under the conditions that it does not undercut `current_block + LOWER_BOUND` or exceed `current_block + UPPER_BOUND`. 
```

File: text/0097-unbonding_queue.md [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+
+
+### Prior Art and References
+- Ethereum proposed a [similar solution](https://blog.stake.fish/ethereum-staking-all-you-need-to-know-about-the-validator-queue/)
+- Alistair did some initial [write-up](https://hackmd.io/SpzFSNeXQM6YScW1iODC_A)
+- There are [other solutions](https://arxiv.org/pdf/2208.05408.pdf) that further mitigate the risk of LRAs.
+
+### The Unresolved Question: Deferred slashing
+
+Currently we defer applying many slashes until 28 days have passed. This was implemented so we can conveniently cancel slashes via governance in the case that the slashing was due to a bug. While rare on Polkadot, such bugs cause a significant fraction of slashes. This includes slashing for attacks other than LRAs for which we've assumed that 2 days is enough to slash. But 2 days in not enough to cancel slashes via OpenGov.
```
Comment (ordian) [inline text/0097-unbonding_queue.md:160]: https://github.com/polkadot-fellows/runtimes/blob/7c69345c75910006d56e60e2f9a93c9d0f44f280/relay/polkadot/src/lib.rs#L654
```suggestion
Currently we defer applying many slashes until 27 days have passed. This was implemented so we can conveniently cancel slashes via governance in the case that the slashing was due to a bug. While rare on Polkadot, such bugs cause a significant fraction of slashes. This includes slashing for attacks other than LRAs for which we've assumed that 2 days is enough to slash. But 2 days in not enough to cancel slashes via OpenGov.
```

File: text/0097-unbonding_queue.md (hunk 9 above)
Comment (jonasW3F) [reply text/0097-unbonding_queue.md:108]: public now

File: text/0097-unbonding_queue.md (hunk 11 above)
Comment (jonasW3F) [reply text/0097-unbonding_queue.md:160]: Right now there is a slash that happened in era 1498 which, polkadot.js apps says, will be applied in era 1526 (28 eras later). Not sure where the discrapency comes from, but since we are talking about days here anyway, I'll just say that it takes *around 28 days*, which should be fine enough.

File: text/0092-unbonding_queue.md (hunk 7 above)
Comment (gpestana) [reply text/0092-unbonding_queue.md]: An even simpler solution would be to clear the unbonding queue and disable the fast unbonding temporarily if there are any pending slashes in the system, and skip calculating the nominators affected by the slash `on-idle` altogether.

> In honest cases, this is a negligible degradation. It attack cases, the worst, everyone is forced to unbond with 28 days.

This remains true.

File: text/0097-unbonding_queue.md [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+## Additional Considerations
+
+### Deferred slashing
+
+Currently we defer applying many slashes until around 28 days have passed. This was implemented so we can conveniently cancel slashes via governance in the case that the slashing was due to a bug. While rare on Polkadot, such bugs cause a significant fraction of slashes. This includes slashing for attacks other than LRAs for which we've assumed that 2 days is enough to slash. But 2 days in not enough to cancel slashes via OpenGov.
+
+Owing to the way exposures, which nominators back validators with how many tokens, are stored, it is hard to search for whether a nominator has deferred slashes that need to be applied to them on chain as of now. So we cannot simply check when a nominator attempts to withdraw their bond. 
+
+There are several ways to solve this issue going forward:
```
Comment (xlc) [inline text/0097-unbonding_queue.md:119]: this need to be decided before vote

Comment (jonasW3F) [reply text/0097-unbonding_queue.md:119]: Do you want me to commit to one solution and remove the other? Or make one more favorable but leave the second for completion?

Comment (jonasW3F) [reply text/0097-unbonding_queue.md:119]: The RFC is implementable and offers two viable solutions

**Review by ordian — APPROVED** [review]

**Review by Ank4n — APPROVED** [review]

**Comment by ggwpez** [comment]
@xlc any more comments? Going to propose soon.

File: text/0097-unbonding_queue.md [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+
+If we detect an LRA of no more than 28 days with the current unbonding period, then we should be able to detect misbehaviour from over 1/3 of validators whose nominators are still bonded. The stake backing these validators is considerable fraction of the total stake (empirically it is 0.287 or so). If we allowed more than this stake to unbond, without checking who it was backing, then the LRA attack might be free of cost for an attacker. The proposed mechansim allows up to half this stake to unbond within 28 days. This halves the amount of tokens that can be slashed, but this is still very high in absolute terms. For example, at the time of writing (19.06.2024) this would translate to around 120 millions DOTs.
+
+Attacks other than an LRA, such as backing incorrect parachain blocks, should be detected and slashed within 2 days. This is why the mechanism has a minimum unbonding period.
```
Comment (eskimor) [inline text/0097-unbonding_queue.md:46]: Ok, I am not to familiar with staking internals, but I will assume that if you unbond, your stake will always remain at least for the duration of the current era. Hypothetically on a network with eras longer than 2 days, it could happen that your stake is securing the network until the end of the era and gets unbonded the very next block. Thus, causing an offense at the end of an era would be risk free?

If that is true, we should bind the unbonding period to multiples of an era, e.g.  2, instead of a fixed 2 days. Also it would be good to measure that period in number of blocks, because then a potential DoS attack would be pointless as it would also increase the unbonding time.

Comment (jonasW3F) [reply text/0097-unbonding_queue.md:46]: The 2 days here is used to make it easier for the reader and putting durations into a more familiar context. The queue itself, however, is specified in blocks in the chapter "Mechanism". Here, `MAX_DURATION` is equal to 403200 blocks and `MIN_DURATION` is equal to 28800. If the queue is implemented on other networks with different era durations, the parameters can easily be adjusted.

Comment (eskimor) [reply text/0097-unbonding_queue.md:46]: Perfect on using blocks! For the relationship to eras: Would be good if that relationship to era duration would be at least documented (not necessarily in this RFC as it is already up for vote), but for sure in the implementation.

Comment (jonasW3F) [reply text/0097-unbonding_queue.md:46]: Makes sense, I'll make to add it to the documentation for the implementation.

**Review by bkchr — APPROVED** [review]

**Comment by ggwpez** [comment]
/rfc process 0xafc7e5a97a974592fd714094c7eb1b23c0f25d43c27b66e639bffebeffb283f9

**Comment by kianenigma** [comment]
Something that I realized a bit late here is how we can support this for pools. Any existing thoughts? 

cc @Ank4n @gpestana.

**Comment by Ank4n** [comment]
> Something that I realized a bit late here is how we can support this for pools. Any existing thoughts?
> 
> cc @Ank4n @gpestana.

We currently track the unlocking of funds for both pool members (SubPools) and staking accounts (UnlockChunks) based on the predetermined era when these funds are set to be unlocked. To align with the RFC, we will probably shift to tracking SubPools and UnlockChunks by the era or block at which the unbonding request is made. At a high level, once the entire UnlockChunk for a specific era is unlocked, the corresponding funds from the SubPool of that era can also be withdrawn. Thoughts?

**Comment by kianenigma** [comment]
Update: This is finally being implemented in https://github.com/paritytech/polkadot-sdk/pull/8298, and I expect it to be either part of the AHM upgrade, or the one right after.

A bit more context: we have now forked `pallet-staking` (used in the RC) and `pallet-staking-async` (will be used in AH post migration of staking there). This is why we closed the initial PR (https://github.com/paritytech/polkadot-sdk/pull/7398) and opened a new one.

The forking and generally interleaving of the implementation with AHM has caused some delay and the need for the team working on AHM (represented by myself) and the implementors (currently BlockDeep) to need to have a few syncs, to ensure the implementation is on the right track and is suitable for the new `staking-async`.

**Comment by Moliholy** [comment]
We have identified a potential gap in the RFC where the rebonding functionality (after the user unbonds) has not been sufficiently described, which is related to #143.

When a user unbonds, it increases the unbonding duration of all stakers that unbond after the operation is performed. When rebonding, the user moves funds from the unbonding queue back to the bonded funds, which has two implications:

### 1. The back of the unbonding queue must be reduced

The RFC assumes that the same amount that was originally unbonded is now rebonded. Based on that assumption, it suggests storing the past delta and simply subtracting that value. However, this assumption is not necessarily valid, as a user may choose an arbitrary amount of stake to rebond, not necessarily equal to the amount unbonded in the past. This fact invalidates the delta that may have been stored.

**Proposed Solution**: The current [implementation](https://github.com/paritytech/polkadot-sdk/pull/8298) recomputes the delta with the current conditions and reduces the back of the unbonding queue with that value.

### 2. Unbonding period compensation issue

All stakers that unbonded between when this unbond-rebond operation took place have their unbond periods artificially high. Ideally, they should see it reduced by the delta described in the previous point. The RFC briefly mentions the existence of an extrinsic for this purpose. However, it is not straightforward to identify existing unbonding operations subject to the corresponding reduction in unbonding time.

**Proposed Solution**: The current [implementation](https://github.com/paritytech/polkadot-sdk/pull/8298) does not propose any solution for now and simply assumes those stakers will not be compensated. Explored solutions include storing and maintaining a list of historical unbonding and rebonding operations performed on the pallet, and attempting to identify for each user the corresponding amount and delta that should be subject to a reduction. However, this approach has a notable computational and storage penalty, so it has not been implemented yet, as further consensus would be needed.
